src/docker.rs

```diff
@@ -90,8 +90,16 @@ pub(crate) enum ClientType {
     SSL,
     #[cfg(windows)]
     NamedPipe,
+    Callback {
+        scheme: String,
+    },
 }
 
+pub type CallbackRetTy =
+    Pin<Box<dyn Future<Output = Result<Response<hyper::body::Incoming>, Error>> + Send>>;
+/// The type of the callback for custom transport.
+pub type CallbackTy = Box<dyn Fn(Request<BodyType>) -> CallbackRetTy + Send + Sync>;
+
```

Comment:
For a more ergonomic API I would suggest using a `trait` instead of the `Fn`, but with default impl for `Fn`

Something like: 

```rs
trait CustomTransport {
  fn request(request: Request<BodyType>) -> ...
}

impl<CallbackTy> CustomTransport for CallbackTy 
where ...
{
    ...
}
```

This way, users can create their own (isolated) structures which may simplify the interaction with API and the same time be more flexible for extensions

WDYT?

Reply:
I added the trait :)

Reply:
Hmm I don't known how to make it work with closures

Reply:
I've left two suggestions which makes it working

Reply:
To make it simpler to check the diff I've opened a PR to your branch: https://github.com/izissise/bollard/pull/1/files